src/web100srv.c

```diff
@@ -1315,15 +1110,66 @@ int run_test(tcp_stat_agent *agent, Connection *ctl, TestOptions *testopt,
              vars.Timeouts, vars.Sndbuf, vars.MaxRwinRcvd, vars.CurrentCwnd,
              mismatch, bad_cable, totalcnt, refresh);
   }
-  if (ctl->ssl != NULL) {
-    SSL_shutdown(ctl->ssl);
-  } else {
-    shutdown(ctl->socket, SHUT_WR);
-  }
-  /* shutdown(ctlsockfd, SHUT_RDWR); */
+  close_connection(ctl);
+  shutdown(ctl->socket, SHUT_WR);
   return (0);
 }
 
+/** Read a single message sent from the parent server process to the child
+ * process, and perform and necessary communication that message necessitates.
+ * @param parent_pipe where to read the message from
+ * @param ctl the Connection to use for the network communication
+ * @param testopt the parameters of the test
+ * @param t_opt the tests the client supports
+ * @return the message value on success, a negative error code on failure.
+ */
+int read_from_parent(int parent_pipe, Connection *ctl, TestOptions *testopt,
```

Comment:
I had a lot of problems related to passing around insufficient context regarding the connections, so in response I have started to strictly prefer passing around a pointer to TestOptions over passing around individual components of TestOptions.

On the other hand, passing around a bool indicating whether or not the client can perform the heartbeat check makes a lot of sense to me, and I have updated the code.